Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -150,72 +156,120 @@ public override bool TryGetVariable(string name, out object value)
+            context.Publish(new CompletionRequestReceived(requestCompletion));
+
+            if (AzShell != null)
+            {
+                // Currently no tab completion when interacting with AzShell.
+                completion = new CompletionRequestCompleted(Array.Empty<CompletionItem>(), requestCompletion);
             }
-            finally
+            else
             {
-                ((PSKernelHostUserInterface)_psHost.UI).ResetProgress();
+                CommandCompletion results = CommandCompletion.CompleteInput(
+                    requestCompletion.Code,
+                    requestCompletion.CursorPosition,
+                    options: null,
+                    _lazyPwsh.Value);
+
+                var completionItems = results.CompletionMatches.Select(
+                    c => new CompletionItem(
+                        displayText: c.CompletionText,
+                        kind: c.ResultType.ToString(),
+                        documentation: c.ToolTip));
+
+                completion = new CompletionRequestCompleted(
+                    completionItems,
+                    requestCompletion,
+                    results.ReplacementIndex,
+                    // The end index is the start index plus the length of the replacement.
+                    results.ReplacementIndex + results.ReplacementLength);
             }
 
+            context.Publish(completion);
             return Task.CompletedTask;
         }
 
-        protected override Task HandleRequestCompletion(
-            RequestCompletion requestCompletion,
-            KernelInvocationContext context)
+        private async Task RunSubmitCodeInAzShell(PowerShell pwsh, string code)
         {
-            var completionRequestReceived = new CompletionRequestReceived(requestCompletion);
-
-            context.Publish(completionRequestReceived);
+            code = code.Trim();
+            bool shouldDispose = false;
 
-            CommandCompletion completion = CommandCompletion.CompleteInput(requestCompletion.Code, requestCompletion.CursorPosition, null, _lazyPwsh.Value);
+            try
+            {
+                if (string.Equals(code, "exit", StringComparison.OrdinalIgnoreCase))
+                {
+                    await AzShell.ExitSession();
+                    shouldDispose = true;
+                }
+                else
+                {
+                    await AzShell.SendCommand(code);
+                }
+            }
+            catch (IOException e)
+            {
+                ReportException(e, pwsh);
+                shouldDispose = true;
+            }
 
-            var completionList = completion.CompletionMatches.Select(c => new CompletionItem(
-                displayText: c.CompletionText,
-                kind: c.ResultType.ToString(),
-                documentation: c.ToolTip
-            ));
+            if (shouldDispose)
+            {
+                AzShell.Dispose();
```

Comment:
You might also want to do `base.RegisterForDisposal` for `AzShell` when it's initialized. For example, do we care if the `PowerShellKernel` gets disposed without someone having done an explicit `exit`?

Reply:
Good point. Will address this.